### PR TITLE
Check implicit receiver in nested initializers

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/32495

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -85,6 +85,84 @@ public override string ToString()
...
+{
+    public B? f;
+    static void Main()
+    {
+        new C() { f = { f2 = null, f3 = null }};
+    }
+}", options: WithNonNullTypesTrue());
+
+            comp.VerifyDiagnostics(
+                // (12,25): warning CS8602: Possible dereference of a null reference.
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Perhaps we should consider putting the warning on the `f =` part, and not the `f2` and `f3` parts. I don't think it's immediately obvious why there's an error on `f2` or `f3`.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: It still won't be obvious, but I adjusted the location a bit ;-)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -85,6 +85,81 @@ public override string ToString()
     }
 }";
 
+        [Fact, WorkItem(32495, "https://github.com/dotnet/roslyn/issues/32495")]
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Please add tests for:
When `f` is both not null and when it's defined in an oblivious context.
Doubly nested (Do we warn on both the nested `f2 = { ... }` and the outer `f = { f2 = { ... } }`?)
Object initializer inside an initializer that returns a nullable reference.

**Comment by 333fred** [comment]
Done review pass (commit 2). Would like some more tests.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I don't understand the last suggestion

---
In reply to: [266132886](https://github.com/dotnet/roslyn/pull/34141#discussion_r266132886) [](ancestors = 266132886)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -85,6 +85,155 @@ public override string ToString()
...
+{
+    public B? f2;
+}
+public class C
+{
+    public C fc;
+    public B? fb;
+    static void Main()
+    {
+        new C() { fc = { fb = { f2 = null} }};
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >fc [](start = 18, length = 2)

What I meant was making `fc` here nullable as well: namely, what will happen when there's a potential nested null derefs.

**Comment by jcouv** [comment]
There was a transient failure in one of the async-iterator tests. I filed issue to investigate further: https://github.com/dotnet/roslyn/issues/34207

**Comment by jcouv** [comment]
@cston @dotnet/roslyn-compiler for a second review.  Thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -85,6 +85,179 @@ public override string ToString()
...
+    {
+        new C() { f = { f2 = null }};
+    }
+}", options: WithNonNullTypesTrue());
+
+            comp.VerifyDiagnostics(
+                // (6,14): warning CS8618: Non-nullable field 'f' is uninitialized.
+                // public class C
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "C").WithArguments("field", "f").WithLocation(6, 14)
+                );
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >); [](start = 16, length = 2)

Shouldn't we report `CS8602: Possible dereference` here too?

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Resolved offline.

**Pull request merged**